server/user_account/permissions.py

```diff
@@ -0,0 +1,23 @@
+from rest_framework.permissions import BasePermission
+from django_server.permissions import IsSystemAdmin, IsCurrentUserTargetUser
+
+
+class UserAccountPermission(BasePermission):
+
+    def has_permission(self, request, view):
+        """
+             Overriding default has_permission method in order to add Custom
+             permissions to our views
+             This can be used either inside the permission_class directly or
+             you can call it from other permission files
+             :param request: Where i retrieve the authenticated user
+             :param view: Where i can retrieve the arguments passed in url, actions ..
+             :return: True/False : Whether the user is allowed to perform CRUD
+        """
+        if view.action in ['retrieve']:
+            return IsCurrentUserTargetUser.has_permission(None, request, view)
+
+        if view.action in ['create']:
+            return IsSystemAdmin.has_permission(None, request, None)
+
+        return False
```

Comment:
So I'll explain the reasoning behind that. Restricting to specific actions is only related to ViewSets if you include it inside the permission class.  Whatever permission is reusable is whatever is under django_server/permissions.py where you can find IsSystemAdmin, IsInventoryManager and IsCurrentUserTargetUser. What you see here is a permission that is custom for the registrationViewSet( renamed  CustomUserView). So here It's not generalized to all the view set because some of them might have different permissions depending on their actions. For now i restricted on this view that only the user can retreive his own info but what i can do is allow also the admin to retreive the info of other people (so i can add that). And for registering a user im just restricting it to admins because we didnt start implementing the other epics where IM can also create accounts. So if we need another custom permission for a different viewset we can just create another one based on our needs, because all the viewsets wont have the same permissions based on their different actions. So we could keep it like that. What do you think?

Reply:
I see what you mean, but I'm always going to question when someone is making something that isn't reusable.

From what I see though, is we're creating another class to simply provide the proper permissions for another class. This couples the 2 classes together (which are in different files) and reduces cohesion for the class that we using these permissions for. I was wondering if we could do something like this:
![viewsetpermissions](https://user-images.githubusercontent.com/40705372/97789612-34380e80-1b98-11eb-9481-fa8c5fb88876.png)

This would allow us to customize the permissions for the viewset, without having to create a new class altogether. The code will be much cleaner this way, as we would have the permissions customized in a method within the class as opposed to a method in another class in another file.

Have you tried something like this?

Reply:
So I've looked into it. It is basically the exact same logic of overriding get_serializer_class() to add different serializers per action which I already made modification for in this commit [2ab2a97](https://github.com/Fruity-Loops/ALTA/pull/116/commits/2ab2a97da2d0420cbd3701c80902f190f54dde0b). So I go proceed and do the same thing with what you mentioned above but I still prefer the original way I did it because it's cleaner I feel. By cleaner, I mean that whatever is related to permissions is just in permissions.py and we can create our custom permissions there without having to write inside the view the code which will end up with a huge viewset. This is my opinion, what do you think.

Reply:
I mean, I get the idea of putting all permissions in one place. 
My concern though is understandability. People looking through the permissions file will see this permission and might wonder why it's implemented like this. Most of the other permissions are implemented in a generic fashion for reusability, except for this one. It only has a single use case for a particular view, which can be confusing for someone who is not familiar with what it's being used for.

In the inverse case, someone might be trying to use this view for some functionality and might not have seen/understood the extra custom class. The result might be someone who gets stuck on this view for hours before understanding some silly permissions that they weren't considering.

I understand the concern for a god class, but I really don't think we've arrived there yet.
I know the issues I brought up sound trivial, as someone who knows how to use their IDE can quickly navigate through the files easily, but to me, it just seems like an unnecessary extra step, and decreases the cohesion of the class. 

If you're still opposed to implementing it though, we can ask for a third opinion.

Reply:
I understand your point, but I still feel it makes the view a bit messier. Also, I implemented the permissions in organization app this way, but still, it can be refactored but I'd be interested what @makchamp thinks since he will be also working intensely on the backend.

Reply:
Hey guys, for this I agree with both of your perspectives. I would say for me it would largely depend on whether or not you think these permission classes will be reused elsewhere in the future. If they will be used in multiple areas in the app then yes it would probably be more responsible to have them abstracted into their own classes. However, if they're strictly related to an appropriate singular viewset, then I would not bother and implement it the way nick described.